extensions/amp-ad-network-doubleclick-impl/0.1/sra-utils.js

```diff
@@ -0,0 +1,292 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {dev} from '../../../src/log';
+import {getEnclosingContainerTypes} from '../../../ads/google/a4a/utils';
+import {
+  isInManualExperiment,
+} from '../../../ads/google/a4a/traffic-experiments';
+
+/**
+ * @const {string}
+ * @visibleForTesting
+ */
+export const TFCD = 'tagForChildDirectedTreatment';
+
+/** @private {!Array<function(!Array<!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl>):?Object<string,string>>} */
+const SRA_JOINERS = [
+  combineInventoryUnits, getCookieOptOut, getAdks, getSizes, getTfcd, isAdTest,
+  getTargetingAndExclusions, getExperimentIds, getIdentity, getForceSafeframe,
+  getPageOffsets, getContainers];
+
+/**
+  * @param {!Array<!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl>} impls
+  * @return {!Object<string, *>}
+  */
+export function constructSRABlockParameters(impls) {
+  const parameters = {'output': 'ldjh', 'impl': 'fifs'};
+  SRA_JOINERS.forEach(joiner => Object.assign(parameters, joiner(impls)));
+  return parameters;
+}
+
+/**
+  * @param {!Array<!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl>} impls
+  * @param {function(!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl):?T} extractFn
+  * @return {?T} value of first instance with non-null/undefined value or null
+  *    if none can be found
+  * @template T
+  * @private
+  */
+function getFirstInstanceValue_(impls, extractFn) {
+  for (let i = 0; i < impls.length; i++) {
+    const val = extractFn(impls[i]);
+    if (val) {
+      return val;
+    }
+  }
+  return null;
+}
+
+/**
+ * Combines inventory unit paths for multiple blocks by building list of
+ * unique path parts in iu_parts and then comma separated list of block
+ * paths using index into iu_parts list.
+ * @param {!Array<!./amp-ad-network-doubleclick-impl.AmpAdNetworkDoubleclickImpl>} impls
+ * @return {?Object<string,string>}
+ * @visibleForTesting
+ */
+export function combineInventoryUnits(impls) {
+  const uniqueIuNames = {};
+  let uniqueIuNamesCount = 0;
```

Comment:
Can remove and replace this with Object.keys(uniqueIuNamesCount).length below

Reply:
That is an O(N) operation (both in memory and runtime) as it requires converting to an array everytime